### ISSUE TITLE
What is the "unit length"? Does gridsize represent the size of a voxel?

### ISSUE THREAD

**Issue by Tyler799** [Forceflow/cuda_voxelizer#14]
I am attempting to create a CPU version of this code. In that effort, I have figured out what part of voxelize_triangle() does, but I need a hand.

What is this "unit length" (unit.x, unit.y, unit.z)? It is the dimensions of the bounding box divided by "gridsize". That would seem to suggest unit length is the number of voxels in each direction, with gridsize being the width of a voxel in each direction. Is that true?

Does that mean this code supports voxels that aren't perfect cubes?

Maybe some of the code has to do with tiling, something I won't need to do and don't totally follow.

Further, I don't follow what the critical point and delta_p are used for. Neither are changed within the voxelize_triangle loop, but both are used. It's very odd.

In any case, here is what I have so far for my rough overview of what the code is doing:

for all triangles...

// fetch the triangle's 9 coordinates (3 vertices)
// compute the edge vectors
// use edge vectors to get tri's normal
// get tri's bounding box in global coordinates (not relative to grid)
// get "t_bbox_grid" (I don't really follow this part. If unit.x is number of voxels in x direction, I don't see what this does. I know what clamp is, though)
// without knowing the prior step, don't want to speculate on the rest

Thanks for any assistance!

**Comment by Forceflow** [comment]
If your bounding box is, for example 4 by 4 by 4 in world units, and you want an 8 by 8 by 8 voxelization, that means the edge length of every voxel will be 0.5., because dividing 4 by 8 is 0.5 :) This means, if you would blow up the voxels to the world coordinate system of the original model, they will be 0.5 length in each direction each.

This unit length is used to convert from world space (whatever coordinate system the original model is in) to voxel coordinates (where the x,y and z axis run from 0 to number_of_voxels).

In a voxel coordinate system, original world vector positions have no meaning anymore. All voxels have coordinates ranging from 0 to number_of_voxels*.

(* well, number of voxels-1, because array indices start at 0, of course :) )

By the way, you can find a CPU version of this voxelization code in my ooc_svo_builder project - several voxelization methods are implemented there: https://github.com/Forceflow/ooc_svo_builder/blob/master/src/svo_builder/voxelizer.cpp (that is old code, but it might help you)

**Comment by Tyler799** [comment]
Thanks for the reply! I follow it conceptually, I just wasn't sure on the variables used. It looks like the older CPU version is extremely similar to what currently exists.

So, in summary:

* gridsize is the dimensions (in voxels) of the 3D grid surrounding the object to be voxelized.
* unit is the size of a voxel in x,y,z

Then why do this:

```
t_bbox_grid.min = glm::clamp(t_bbox_world.min / info.unit, glm::vec3(0.0f, 0.0f, 0.0f), gridsize - 1);
```

If you divide the position of the bounding box by the size of a voxel, what does that represent? If the bounding box around the object started at 2,2,0 , and your voxels are 0.5 wide in each direction, you now get t_bbox_grid_min = 4, 4, 0. Does that mean that effectively the minimum position is is _as if the entire scene was voxelized_ and not just the object? So it's relative to true 0,0,0?



One other question, I noticed that t_bbox_grid is integer, not float. But I don't see it explicitly converted. Does glm::clamp do that implicitly? Since t_bbox_world is a float, and unit can be a float, the result of their division is a float. Or is it just because it's being stored in an ```ivec3```. I guess that must be it, though I didn't see any warnings about it when compiling.

Also, why have ```delta_p``` when it is just equal to ```info.unit```?

Thanks again for all the assistance!

**Comment by Forceflow** [comment]
Correct about gridsize: I always make sure it's as tight as possible around the model, to preserve the best detail.

I think the code you are referring to tries to find out what the lowest _possible_ voxel coordinate is (in 1 dimension) that a _triangle_ (hence the t_ prefix :p) could _possibly touch_. That code then proceeds to (quite dumb) test all the possible voxels that a triangle might touch against the actual triangle using standard triangle/box intersection methods. So this is a small part of testing 1 triangle, not the whole model.

I think it indeed is getting rounded down automatically because it's being stored in an ivec3. Since C (and C++) always rounds down to the lowest integer, that is correct. Even 1.99 will round down to 1.

**Comment by Tyler799** [comment]
Thanks again. What about ```delta_p``` and ```info.unit```? Why have delta_p if it is just equal to unit and is not changed in the loop? Also, just confirming since you haven't mentioned yet - unit is the width of a voxel in each direction, right?

**Comment by Forceflow** [comment]
What file are you specifically referring to? The one in ooc_svo_builder or here?

**Comment by Tyler799** [comment]
My bad, I meant the one in this repository.

As well, I wanted to check - this algorithm does work with boundary representations of shapes that are hollow, right? (Assuming tit is a well-formed boundary representation) IE, I have an STL file I need to voxelize that is a hollow cylinder.

And no additional transformation or translation of the vertices, tris, etc is needed before running the ```voxelize_triangle``` method, right? 

I'm getting some very strange results. I'm printing out x,y slices of the voxels to the screen representing this cylinder: https://www.thingiverse.com/thing:1740003. It's aligned down the Z axis.

However, I get results like this:

https://pastebin.com/wd1i63kC

The bottom right and left corners are correctly not turned on, but the inside that is hollow and upper right/left are on, which shouldn't happen. As I go down slices, it shifts to have zeros on the top right/left instead.

I've been pouring over why this might be happening. My code is almost a carbon-copy of the one you have for that method, just using some nanort types instead of glm types. 

And, since our voxels are always cubes, I can just do:

```nanort::real3<float> unit(voxelWidth, voxelWidth, voxelWidth);```

Other than that, they're nearly the same. I don't use any Morton encoding anywhere, but I don't think that's required by the implementation, is it?

**Comment by Tyler799** [comment]
I'm still getting very odd results - but I just realized something. Does that mean that this code only does a *surface* voxelization, not a *solid* voxelization? That isn't why this code isn't working for me, but if it's true then I've unfortunately wasted a ton of time.

**Comment by Tyler799** [comment]
I have two good pieces of news.

First, I figured out my issue with the surface voxelization. It was dumb and entirely my fault. My implementation of ```clamp``` was making a local copy instead of taking a reference, and my method for flattening the 3D index had a typo. 

```position = x + (y * gridwidth.x) + (z * part->gridwidth.x * gridwidth.y);```

Here are the fixed the results for a box showing the first 2 Z-slices: https://pastebin.com/C16NDTF4

And the results for the cylinder aligned down the Z-axis: https://pastebin.com/251y00Hh

Both are absolutely perfect. (Minus the fact that the grid is low-res intentionally)

As well, while this is just a surface voxelization, I found a solid voxelization example in "Practical Binary Surface and Solid Voxelization with Direct3D 11" in the book _GPU Pro 3_. I don't own it, but the paper's author gave me a copy of the code used in it. Between that and your code, along with the fixes I found, I should be able to get it working.

Thanks for all the assistance!
